corgi/tasks/brew.py

```diff
@@ -515,29 +532,21 @@ def load_brew_tags(build_id: str, brew_tags: list[str]) -> int:
         for tag in distinct_stream_tags:
             brew, build_type, _ = _relation_context_for_stream(stream_name)
             logger.info(f"Creating relations for {stream_name} and {tag}")
-            no_created += _create_relations_for_tag((build_id,), tag, build_type, stream_name, None)
+            # We do this inline instead of via create_relations function because
+            # it's the only time we call it where we have a software_build object created
+            _, created = ProductComponentRelation.objects.get_or_create(
+                external_system_id=tag,
+                product_ref=stream_name,
+                build_id=software_build.build_id,
+                build_type=build_type,
+                software_build=software_build,
+                defaults={"type": ProductComponentRelation.Type.BREW_TAG},
+            )
```

Comment:
get_or_create arguments should only use fields that appear in a unique_together constraint, in order to guarantee it behaves atomically. In this case, I think that's external_system_id, product_ref, and build_id. Maybe build_type was added to the constraint as well.

If build_type isn't part of the unique_together constraint, it should be added, or we should move all the build_type values into the defaults dict.

software_build should also be part of the defaults dict, and we should probably switch this to use update_or_create instead.

Sorry for completely forgetting about this until just now. Can you please check the rest of your changes to make sure all other get_or_create / update_or_create usages are atomic?

```suggestion
            _, created = ProductComponentRelation.objects.update_or_create(
                external_system_id=tag,
                product_ref=stream_name,
                build_id=software_build.build_id,
                defaults={
                    "type": ProductComponentRelation.Type.BREW_TAG
                    "build_type": build_type,
                    "software_build": software_build,
                },
            )
```

Reply:
Build_type is part of the unique_together constraint. I moved software_build to defaults dict thought, thanks for noticing that.